design/ceph/osd-migration.md

```diff
@@ -0,0 +1,227 @@
+
+### Existing OSDs
+
+- After upgrading an existing rook ceph cluster to a version that supports `bluestore-rdr`, users can upgrade the `bluestore` based OSDs to use `bluestore-rdr`
+- Backend store of the OSDs can not be overridden, so this update will require the OSDs to be replaced one by one.
+- In order to start this osd migration, users need to patch the ceph cluster spec like below:
+
+```yaml
+storage:
+  osdStore:
+    type: bluestore-rdr
+    updateStore: yes-really-update-store
+```
+
+#### API
+
+```go
+// OSDReplaceInfo represents an OSD that needs to replaced
+type OSDReplaceInfo struct {
+	ID   int    `json:"id"`
+	Path string `json:"path"`
+	Node string `json:"node"`
+}
+
+// ReplaceQueue contains the list of OSDs to be replaced one at a time
+type ReplaceQueue struct {
+	OSDs []OSDReplaceInfo
+}
+
+// OSDReplaceConfig contains the status of the OSD replacement
+type OSDReplaceConfig struct {
+	OSDList              *ReplaceQueue   `json:"osdList,omitempty"`
+	IsPgHealthy          bool            `json:"isPgHealthy,omitempty"`
+	PgHealthMessage      string          `json:"pgHealthMessage,omitempty"`
+	LastReplacedOSD      *OSDReplaceInfo `json:"lastReplacedOSD,omitempty"`
+	LastReplaceTimeStamp time.Time       `json:"lastReplacedTimeStamp,omitempty"`
+	StartTimeStamp       time.Time       `json:"startTimeStamp,omitempty"`
+}
+```
+
+**OSDReplaceInfo**: Represents an OSD that needs to be replaced
+  - **ID**: OSD ID of the OSD to be replaced.
+  - **Path**: Disk or pvc claim that the OSD is consuming.
+  - **Node**: Node on which the OSD is running
+
+
+**ReplaceQueue**: This is a queue implementation containing a list of OSDs to be replaced. It will support basic queue operations like push, pop, size etc.
+
+
+**OSDReplaceConfig**: This will hold the current status/progress of the OSD replacement in the cluster.
+  - **OSDList**:  List of OSDs that are yet to be replaced.
+  - **IsPgHealth**: Whether Placement Groups are healthy or not. This is a boolean field.
+  - **PgHealthMessage**: This will contain a detailed health message of the placement groups. For example: `active+clean`
+  - **LastReplaceOSD**: This field will contain details of the OSD that was last replaced.
+  - **LastReplaceTimeStamp**: Time at which the last OSD was replaced.
+  - **StartTimeStamp**: Time at which the whole OSD replacement process was started. This will help users know how much time has elapsed since the first OSD was replaced.
+
+
+#### Replace OSD :
+
+Operator reconciler will perform following sequence of steps to replace an OSD:
+  1. List out all the OSDs that need to be replaced. This list will contain OSDs that have `osd-store:<osd store type>` pod label not matching the osd store type provided in `spec.storage.osdStore.type`. This list will be stored in `OSDReplaceConfig.OSDList` and will be used a queue.
+  2. If the PG status is not `active+clean`, don't pop and OSD from the `OSDReplaceConfig.OSDList`
```

Comment:
I think we might want to have a step between 1 and 2 that is like: 
"If the CephCluster storage config does not allow replacing OSDs for updates, emit a warning event and error message in Rook's log, then proceed with standard (no replacement) OSD update logic"

Reply:
What situation will lead to - `CephCluster storage config does not allow replacing OSDs for updates`, 
other than PGs are not active due to last OSD replacement?